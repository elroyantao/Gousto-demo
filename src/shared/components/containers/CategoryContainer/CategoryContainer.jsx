import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Helmet from 'react-helmet'

import ProductList from '../../presentation/ProductList/ProductList'
import Search from '../../presentation/Search/Search'
import NoItemFound from '../../presentation/NoItemFound/NoItemFound'

import { setSearchTerm, clearSearchTerm } from '../../../actions/searchTermAction'

@connect((state) => ({
  categories: state.categories,
  products: state.products,
  searchTerm: state.searchTerm
}), { setSearchTerm, clearSearchTerm })
export default class CategoryContainer extends Component {
  static propTypes = {
    categories: PropTypes.arrayOf(PropTypes.shape({
      box_limit: PropTypes.number,
      hidden: PropTypes.bool,
      id: PropTypes.string,
      is_default: PropTypes.bool,
      recently_added: PropTypes.bool,
      title: PropTypes.string
    })),
    match: PropTypes.object,
    products: PropTypes.array,
    searchTerm: PropTypes.string,
    setSearchTerm: PropTypes.func,
    clearSearchTerm: PropTypes.func
  }

  static defaultProps = {
    categories: [],
    products: [],
    searchTerm: ''
  }

  componentWillUpdate({ match: { params: { categoryId: newCategoryId } } }) {
    const { match: { params: { categoryId } }, clearSearchTerm } = this.props
    if (categoryId !== newCategoryId) clearSearchTerm()
  }

  filterProducts() {
    const { match, products, searchTerm } = this.props
    const search = searchTerm.toLowerCase()

    return products.filter((product) => {
      return this.matchSearch(product, search) &&
        product.categories.some((category) => {
          return category.id === match.params.categoryId
        })
    })
  }

  matchSearch(product, search) {
    return product.title.toLowerCase().includes(search)
      || product.description.toLowerCase().includes(search)
      || search === ''
  }

  render() {
    const products = this.filterProducts()
    const { categories, match, searchTerm, setSearchTerm } = this.props
    const selectedCategory = categories && categories.find((category) =>
      category.id === match.params.categoryId
    )
    const name = selectedCategory && selectedCategory.title
    return (
      <div>
        <Helmet title={`Gousto - Category : ${name}`} />
        <Search searchTerm={searchTerm} onSearch={setSearchTerm} />
        { products.length ?
          <ProductList products={products} /> :
          <NoItemFound />
        }
      </div>
    )
  }
}
