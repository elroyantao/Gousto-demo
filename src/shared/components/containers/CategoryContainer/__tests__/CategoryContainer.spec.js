import testComponentHelper from '../../../../lib/test-helper'
import CategoryContainer from '../CategoryContainer'

describe('<CategoryContainer/>', () => {
  const renderComponent = testComponentHelper(CategoryContainer.WrappedComponent)
  const initialProps = {
    products: [
      {
        id: '1',
        title: 'title 1',
        description: 'description 1',
        categories: [{ id: 'fake-id' }]
      },
      {
        id: '2',
        title: 'title 2',
        description: 'description 2 ',
        categories: [{ id: 'fake-id-diff' }]
      },
      {
        id: '3',
        title: 'title 3 search tItle',
        description: 'description 3',
        categories: [{ id: 'fake-id' }]
      },
      {
        id: '4',
        title: 'title 4',
        description: 'description 4 search description',
        categories: [{ id: 'fake-id' }]
      }
    ],
    categories: [
      {
        id: 'fake-id',
        title: 'one'
      },
      {
        id: 'fake-id-diff',
        title: 'two'
      },
      {
        id: 'fake-id-no-products',
        title: 'No Products'
      }
    ],
    searchTerm: '',
    match: {
      params: { categoryId: 'fake-id' }
    },
    setSearchTerm: jest.fn(),
    clearSearchTerm: jest.fn()
  }

  describe('@renders', () => {
    it('in default state', () => {
      const { wrapper, getTree } = renderComponent(initialProps)
      const productsShown = wrapper.find('ProductList').props().products
      expect(productsShown.length).toBe(3)
      expect(getTree())
        .toMatchSnapshot()
    })
    it('when no products for the matched category', () => {
      const { getTree } = renderComponent({
        ...initialProps,
        match: {
          params: { categoryId: 'fake-id-no-products' }
        }
      })
      expect(getTree())
        .toMatchSnapshot()
    })
    it('when 1 products matches category and search matches title', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialProps,
        searchTerm: 'search title'
      })
      const productsShown = wrapper.find('ProductList').props().products
      expect(productsShown.length).toBe(1)
      expect(productsShown[0].id).toBe('3')
      expect(getTree())
        .toMatchSnapshot()
    })
    it('when 1 products matches category and search matches description', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialProps,
        searchTerm: 'search dEscription'
      })
      const productsShown = wrapper.find('ProductList').props().products
      expect(productsShown.length).toBe(1)
      expect(productsShown[0].id).toBe('4')
      expect(getTree())
        .toMatchSnapshot()
    })
  })
  describe('@lifecycle', () => {
    describe('componentWillUpdate', () => {
      let renderedComponent
      beforeAll(() => {
        jest.clearAllMocks()
        renderedComponent = renderComponent(initialProps)
      })
      it('should not call clearSearchTerm when the categoryId does not change', () => {
        const { instance, wrapper } = renderedComponent
        expect(instance.props.clearSearchTerm).not.toHaveBeenCalled()
        wrapper.setProps({ searchTerm: 'is changed' })
        expect(instance.props.clearSearchTerm).not.toHaveBeenCalled()
      })
      it('should call clearSearchTerm when the categoryId change', () => {
        const { instance, wrapper } = renderedComponent
        expect(instance.props.clearSearchTerm).not.toHaveBeenCalled()
        wrapper.setProps({ match: { params: { categoryId: 'will-change' } } })
        expect(instance.props.clearSearchTerm).toHaveBeenCalledTimes(1)
      })
    })
  })
  describe('@events', () => {
    describe('onSearch search', () => {
      it('should call setSearchTerm', () => {
        const searchTerm = 'new search term'
        const { instance, wrapper } = renderComponent(initialProps)
        expect(instance.props.setSearchTerm).not.toHaveBeenCalled()
        wrapper.find('Search').simulate('search', searchTerm)
        expect(instance.props.setSearchTerm).toHaveBeenCalledTimes(1)
        expect(instance.props.setSearchTerm).toHaveBeenLastCalledWith(searchTerm)
      })
    })
  })
})
