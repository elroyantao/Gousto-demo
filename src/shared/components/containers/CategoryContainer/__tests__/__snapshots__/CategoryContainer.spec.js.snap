// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryContainer/> @renders in default state 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="Gousto - Category : one"
  />
  <Search
    onSearch={[Function]}
    searchTerm=""
  />
  <ProductList
    products={
      Array [
        Object {
          "categories": Array [
            Object {
              "id": "fake-id",
            },
          ],
          "description": "description 1",
          "id": "1",
          "title": "title 1",
        },
        Object {
          "categories": Array [
            Object {
              "id": "fake-id",
            },
          ],
          "description": "description 3",
          "id": "3",
          "title": "title 3 search tItle",
        },
        Object {
          "categories": Array [
            Object {
              "id": "fake-id",
            },
          ],
          "description": "description 4 search description",
          "id": "4",
          "title": "title 4",
        },
      ]
    }
  />
</div>
`;

exports[`<CategoryContainer/> @renders when 1 products matches category and search matches description 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="Gousto - Category : one"
  />
  <Search
    onSearch={[Function]}
    searchTerm="search dEscription"
  />
  <ProductList
    products={
      Array [
        Object {
          "categories": Array [
            Object {
              "id": "fake-id",
            },
          ],
          "description": "description 4 search description",
          "id": "4",
          "title": "title 4",
        },
      ]
    }
  />
</div>
`;

exports[`<CategoryContainer/> @renders when 1 products matches category and search matches title 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="Gousto - Category : one"
  />
  <Search
    onSearch={[Function]}
    searchTerm="search title"
  />
  <ProductList
    products={
      Array [
        Object {
          "categories": Array [
            Object {
              "id": "fake-id",
            },
          ],
          "description": "description 3",
          "id": "3",
          "title": "title 3 search tItle",
        },
      ]
    }
  />
</div>
`;

exports[`<CategoryContainer/> @renders when no products for the matched category 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="Gousto - Category : No Products"
  />
  <Search
    onSearch={[Function]}
    searchTerm=""
  />
  <NoItemFound />
</div>
`;
